---
title: "Taylors_Law_Chapter_01"
author: "Soren Maret"
date: "04.02.2025"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyr)
library(tidyverse)
library(dplyr)
library(ggplot2)
```
##GOAL##: Do a proof of concept that their is a scaling (power law) relationship in the gene expression pattern of dox exposed RNA-seq data.

Specifically I will investigate if Taylors Law, an imperical law from ecology, which explains the scaling relationship between the mean and standard deviation presents in our data. 

Taylors law is expressed in the following way: 

sd = b * mean^a

This related the standard deviation (stdev) to the mean by 2 constants b and a where a is our scaling value and b is a normaling constant 

We can rearrange this to using logarithms to get these values by themselves. to do this, since our bases are not equal, we will use the natural logarithm (ln):

ln(sd) = a * ln(mean) + ln(b)

Notice how this is the equation of a line. we can derive the values for a (slope) and b (y-intercept), which are the biologically informative ones, by running a linear regression with high enough statistical signifigance (measured by the regression constant R)

for further reading see the resources below:
https://necsi.edu/power-law
https://www.nature.com/articles/s41562-016-0012
https://journals.aps.org/prx/abstract/10.1103/PhysRevX.12.011051#:~:text=Popular%20Summary,varies%20from%20cell%20to%20cell.

In our case we will examine the mean expression of a gene compaired to its standard deviation. to do this we will use our counts matrix produced from our nextflow run.

Lets load in some data shall we? To do a proof of concept I will run both human and mouse data in parallel.
```{r reading in data and creating files for downstream analysis}
Human_counts <- read_tsv("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/HUMAN/salmon.merged.gene_counts.tsv")

Mouse_counts <- read_tsv("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/MOUSE/salmon.merged.gene_counts.tsv")

#cool! now lets remove the logical values to our results dataframe which we ill call taylor_values_human or taylor_values_mouse

#lets do this for humans:
#making a dataframe that contains our "gene_id" collumn:
taylor_values_human <- Human_counts["gene_id"] 
#now lets move gene_names over as well
taylor_values_human <- cbind(taylor_values_human, Human_counts[c("gene_name")])
#looks just like our g2s.df from earlier this semester!

#now for the mice: 
taylor_values_mouse <- Mouse_counts["gene_id"] 
#now lets move gene_names over as well
taylor_values_mouse <- cbind(taylor_values_mouse, Mouse_counts[c("gene_name")])
#great! this all worked out nicely!
```
Now we have our new dataframes which we can load all of our means and standard deviations into so we can later run analysis. in order to successfully do these calculations we need to remove the charaters and make our dataframes numeric.We can do this in any order. 
```{r Making data numeric and moving character based collumns to our taylor_values sheet}
#first, lets remove the collumns gene_name and gene_id as they are characters and will just become N/As
#removes gene_id
Human_counts <- Human_counts[, !(names(Human_counts)) %in% "gene_id"]
#removes gene_name
Human_counts <- Human_counts[, !(names(Human_counts)) %in% "gene_name"]
#makes out counts matrix numeric
Human_counts <- data.frame(lapply(Human_counts, as.numeric))

#we will do the same for our mouse data:
Mouse_counts <- Mouse_counts[, !(names(Mouse_counts)) %in% "gene_id"]
#removes gene_name
Mouse_counts <- Mouse_counts[, !(names(Mouse_counts)) %in% "gene_name"]
#makes out counts matrix numeric
Mouse_counts <- data.frame(lapply(Mouse_counts, as.numeric))
```
Great! all is well with the cosmos. Now Taylors Law looks at 2 things an independent variable: the mean of a gene (in our case, often its a taxa) and an indepdent variable: the standard deviation of that gene. so we need to calculate these values as values in our enviornment. 
```{r taking the means and standard deviations + plotting}
#here is the anaylsis for the human counts: 
means_human <- rowMeans(Human_counts)
#now for the standard deviation accross the rows: 
sd_human <- apply(Human_counts, 1, sd)

#now we can do the same for our mouse data
means_mouse <- rowMeans(Mouse_counts)
#now for the standard deviation accross the rows: 
sd_mouse <- apply(Mouse_counts, 1, sd)

#Sweet!! now lets plot this just to see whats up?
plot(means_human, sd_human)
plot(means_mouse, sd_mouse)
#cool! our data is linear but the values varie massively! but on the bright side this is a proof of concept!!!
```
##RESULT##: 
1. our RNA-seq data is explained by taylors law.
2. most of the genes are expressed at a low level, with fewer and fewer genes being expressed at very high levels (orders of magnitude more)
3. the human and mouse data have different slopes (values of coefficient "a".

Now lets move onto making this all analyzable in chapter 02: but first, some housekeeping:
```{r Updating Dataframes and saving results}
taylor_values_human <- cbind(taylor_values_human, mean = means_human, sd = sd_human)

taylor_values_mouse <- cbind(taylor_values_mouse, mean = means_mouse, sd = sd_mouse)

save(Human_counts, Mouse_counts, taylor_values_human, taylor_values_mouse, means_human, means_mouse, sd_human, sd_mouse, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/TPL_ANALYSIS/RData/Chapter_01.RData")
```