---
title: "Class Project"
author: "ML"
date: "2025_03_31"
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
library(tidyverse)
library(pheatmap)
library(DESeq2)
```

```{r MS data}
#Raw counts file
MS_counts <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/MOUSE/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

#make g2s
g2s <- data.frame(
  gene_id = rownames(MS_counts),
  gene_name = MS_counts[, 1]
)
#Remove gene_name
MS_counts <- MS_counts[, -1]

# Round counts to integer mode required for DESEQ2
MS_integer <- round(MS_counts)

col_data <- data.frame(
  sample_id = colnames(MS_counts))

split_values <- strsplit(col_data$sample_id, "_")

# So here we will go through each row of split_values and run a "generic function(x)" 
# We will then retain the second item which is the time point value in sample_id
condition_values <- sapply(split_values, function(x) x[[1]])
time_values <- sapply(split_values, function(x) x[[2]])

# Similar to above we are using apply to grab the third fragment in split_values (replicate value)
replicate_values <- sapply(split_values, function(x) x[[3]])

# Adding condition and time point into samplesheet for DESEQ2
col_data$time <- time_values
col_data$condition <- condition_values
# Now let's add another column for replicate
col_data$replicate <- replicate_values

#facotring timepoint column
col_data$time <- factor(col_data$time, levels = c("0", "12", "24", "48", "96"))
levels(col_data$time)
```

```{r MS DESeq2}
stopifnot(all(colnames(MS_counts) == col_data$sample_id))

# Create the DESeqDataSet
MS_dds <- DESeqDataSetFromMatrix(countData = MS_integer, colData = col_data, design = ~ time)
# Perform the LRT to find genes changing due to MS
MS_dds <- DESeq(MS_dds, test = "LRT", reduced = ~ 1)

MS_dds <- MS_dds [  rowSums ( counts (MS_dds) ) >  1 , ] 
nrow (MS_dds)

#Curating all results into data frame
resultsNames(MS_dds)

result_names <- resultsNames(MS_dds)

results_names <- result_names[-1]

res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# For loop to get all results per time point  

for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(MS_dds, name = results_name)
  tmp_res_df <- res %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  res_df <- dplyr::bind_rows(res_df, tmp_res_df)
  
}

MS_res_df <- res_df

save(MS_dds, MS_res_df, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/MS_LRT_res_df.RData")
```

```{r MS sig gene}
# Calculate the maximum fold-change in any one timepoint
  MS_maxfc <- MS_res_df %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  # merge max shrnklfc into dataframe
  MS_res_df <- MS_res_df %>%
    left_join(MS_maxfc)
  
  MS_res_df_padj0.05 <- MS_res_df %>% 
  filter(padj <= 0.05)
  print(length(unique(MS_res_df_padj0.05$gene_id)))#7665
  
  MS_sig <- MS_res_df_padj0.05 %>%
  filter(max_fc >= 1)
  print(length(unique(MS_sig$gene_id)))#609
  
  #compare with non LRT results
  load("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/soma3841/MASTER_CLASS/lessons/06_Differential_expression_analyses/results/DESEQ_results.rdata")
  filtered_res_df <- filtered_res_df %>%
  filter(!(gene_name %in% MS_sig$gene_name))
  print(length(unique(filtered_res_df$gene_id))) #135
  
  MS_filtered_res_df <- filtered_res_df
  save(MS_sig, MS_filtered_res_df, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/MS_sig.RData")
```

```{r HS data}
#Raw counts file
human_counts <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/HUMAN/salmon.merged.gene_counts.tsv", header=TRUE, row.names=1)

human_counts <- human_counts %>%
  select(gene_name, gfp_0_1, gfp_0_2, gfp_0_3, gfp_12_1, gfp_12_2, gfp_12_3, gfp_24_1, gfp_24_2, gfp_24_3, gfp_48_1, gfp_48_2, gfp_48_3, gfp_96_1, gfp_96_2, gfp_96_3)

#make g2s
g2s <- data.frame(
  gene_id = rownames(human_counts),
  gene_name = human_counts[, 1]
)
#Remove gene_name
human_counts <- human_counts[, -1]

# Round counts to integer mode required for DESEQ2
human_integer <- round(human_counts)

col_data <- data.frame(
  sample_id = colnames(human_counts))

split_values <- strsplit(col_data$sample_id, "_")

# So here we will go through each row of split_values and run a "generic function(x)" 
# We will then retain the second item which is the time point value in sample_id
condition_values <- sapply(split_values, function(x) x[[1]])
time_values <- sapply(split_values, function(x) x[[2]])

# Similar to above we are using apply to grab the third fragment in split_values (replicate value)
replicate_values <- sapply(split_values, function(x) x[[3]])

# Adding condition and time point into samplesheet for DESEQ2
col_data$time <- time_values
col_data$condition <- condition_values
# Now let's add another column for replicate
col_data$replicate <- replicate_values

#facotring timepoint column
col_data$time <- factor(col_data$time, levels = c("0", "12", "24", "48", "96"))
levels(col_data$time)
```

```{r human DESeq2}
stopifnot(all(colnames(human_counts) == col_data$sample_id))

# Create the DESeqDataSet
human_dds <- DESeqDataSetFromMatrix(countData = human_integer, colData = col_data, design = ~ time)
# Perform the LRT to find genes changing due to human
human_dds <- DESeq(human_dds, test = "LRT", reduced = ~ 1)

human_dds <- human_dds [  rowSums ( counts (human_dds) ) >  1 , ] 
nrow (human_dds)

#Curating all results into data frame
resultsNames(human_dds)

result_names <- resultsNames(human_dds)

results_names <- result_names[-1]

res_df <- data.frame("gene_id" = character(), 
                     "baseMean" = numeric(), 
                     "log2FoldChange" = numeric(), 
                     "lfcSE" = numeric(),
                     "stat" = numeric(),
                     "pvalue" = numeric(),
                     "padj" = numeric(),
                     "gene_name" = character(),
                     "result_name" = character())

# For loop to get all results per time point  

for(i in 1:length(results_names)) {
  results_name <- results_names[i]
  res <- results(human_dds, name = results_name)
  tmp_res_df <- res %>% as.data.frame() %>%
    rownames_to_column("gene_id") %>%
    merge(g2s) %>%
    mutate(result_name = results_name)
  res_df <- dplyr::bind_rows(res_df, tmp_res_df)
  
}

human_res_df <- res_df

save(human_dds, human_res_df, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/human_LRT_res_df.RData")
```

```{r human sig gene}
# Calculate the maximum fold-change in any one timepoint
  human_maxfc <- human_res_df %>%
    group_by(gene_id) %>%
    summarize(max_fc = max(abs(log2FoldChange))) 
  
  # merge max shrnklfc into dataframe
  human_res_df <- human_res_df %>%
    left_join(human_maxfc)
  
  human_res_df_padj0.05 <- human_res_df %>% 
  filter(padj <= 0.05)
  print(length(unique(human_res_df_padj0.05$gene_id)))#11116
  
  human_sig <- human_res_df_padj0.05 %>%
  filter(max_fc >= 1)
  print(length(unique(human_sig$gene_id)))#1937
  
  #compare with non LRT results
  load("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/HUMAN/time_point_res_df.RData")
  
  filtered_res_df <- res_df %>%
  filter(padj < 0.05, abs(log2FoldChange) > 1)
  print(length(unique(filtered_res_df$gene_id)))#2078
  
  filtered_res_df <- filtered_res_df %>%
  filter(!(gene_name %in% human_sig$gene_name))
  print(length(unique(filtered_res_df$gene_id)))#330
  
  human_filtered_res_df <- filtered_res_df
  save(human_sig, human_filtered_res_df, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/human_sig.RData")
```


## [2024 Dataset](https://doi.org/10.1016/j.celrep.2015.04.023) 
```{r syntenic gene (Ulitsky)}
gencode_synteny <- read.delim("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/GENCODEvGENCODE_orthologs_v3.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")

gencode_synteny_filtered <- gencode_synteny %>% 
  filter(SpeciesWithSynteny == "[Mouse]")

gencode_synteny_filtered <- gencode_synteny_filtered %>% 
  mutate(across(everything(), ~ str_replace_all(.x, "\\[|\\]", "")))

gencode_synteny_filtered <- gencode_synteny_filtered %>%
  separate_rows(Mouse.Synteny, sep = ", ")

gencode_synteny_filtered <- gencode_synteny_filtered %>%
  mutate(across(c(GeneId, Mouse.Synteny), ~ str_replace(.x, "\\..*", "")))

#In gencode_synteny_filtered, "Mouse.Synteny" is transcript ID, need annotate with gene ID
library(biomaRt)

# Connect to Ensembl database
mart <- useMart("ensembl", dataset = "mmusculus_gene_ensembl")

# Retrieve transcript-to-gene mapping
tx2gene <- getBM(attributes = c("ensembl_transcript_id", "ensembl_gene_id"),
                 mart = mart)

gencode_synteny_filtered <- gencode_synteny_filtered %>%
  left_join(tx2gene, by = c("Mouse.Synteny" = "ensembl_transcript_id"))

#transform to stable gene id
MS_sig <- MS_sig %>%
  mutate(across(gene_id, ~ str_replace(.x, "\\..*", "")))
human_sig <- human_sig %>%
  mutate(across(gene_id, ~ str_replace(.x, "\\..*", "")))

common_sig <- gencode_synteny_filtered %>%
  filter(GeneId %in% human_sig$gene_id,
         ensembl_gene_id %in% MS_sig$gene_id)

common_sig_filtered <- common_sig %>%
  dplyr::select(GeneId, ensembl_gene_id)

human_common_sig <- human_sig %>%
  filter(gene_id %in% common_sig$GeneId)
human_common_sig <- human_common_sig %>%
  left_join(common_sig_filtered, by = c("gene_id" = "GeneId"))

MS_common_sig <- MS_sig %>%
  filter(gene_id %in% common_sig$ensembl_gene_id)
MS_common_sig <- MS_common_sig %>%
  left_join(common_sig_filtered, by = c("gene_id" = "ensembl_gene_id"))
MS_common_sig <- MS_common_sig %>%
  dplyr::rename(ensembl_gene_id = GeneId)
#5

save(tx2gene, gencode_synteny_filtered, human_common_sig, MS_common_sig, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/gencode_synteny_filtered.RData")
```


## Using [2015 Dataset](https://doi.org/10.1038/s41588-023-01620-7)
Here, we use the comparison from [Hezroni et al](https://doi.org/10.1038/s41588-023-01620-7) to determine which genes are conserved between the human and mouse genes.
```{r syntenic lncRNA gene (fish paper)}
synteny <- read.delim("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/human_mouse_syntenic_lncRNA.txt", header = TRUE, stringsAsFactors = FALSE, sep = "\t")
#data already tidy

common_sig <- synteny %>%
  filter(human.gene %in% human_sig$gene_id,
         mouse.gene %in% MS_sig$gene_id)

common_sig_filtered <- common_sig %>%
  dplyr::select(human.gene, mouse.gene)

human_common_sig_lncrna <- human_sig %>%
  filter(gene_id %in% common_sig$human.gene)
human_common_sig_lncrna <- human_common_sig_lncrna %>%
  left_join(common_sig_filtered, by = c("gene_id" = "human.gene"))
human_common_sig_lncrna <- human_common_sig_lncrna %>%
  dplyr::rename(ensembl_gene_id = mouse.gene)

MS_common_sig_lncrna <- MS_sig %>%
  filter(gene_id %in% common_sig$mouse.gene)
MS_common_sig_lncrna <- MS_common_sig_lncrna %>%
  left_join(common_sig_filtered, by = c("gene_id" = "mouse.gene"))
MS_common_sig_lncrna <- MS_common_sig_lncrna %>%
  dplyr::rename(ensembl_gene_id = human.gene)
#5

save(synteny, common_sig, human_common_sig_lncrna, MS_common_sig_lncrna, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/common_lncRNA.RData")
```

## Using bioMart
Finally, we used the bioMart function in the [Ensemble gene browser](http://www.ensembl.org/biomart/martview/7a6e3f7aaa7334709b8e81ada35342e4) to retrieve orthologs between the human and mouse genes.
```{r conserved protein coding gene (bioMart)}
orthologs <- read.csv("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/mart_export.txt", header = TRUE, stringsAsFactors = FALSE, sep = ",")

orthologs_filtered <- orthologs[orthologs$Mouse.homology.type == "ortholog_one2one", ]

common_orthologs <- orthologs_filtered %>%
  filter(Gene.stable.ID %in% human_sig$gene_id,
         Mouse.gene.stable.ID %in% MS_sig$gene_id)

common_orthologs_filtered <- common_orthologs %>%
  dplyr::select(Gene.stable.ID, Mouse.gene.stable.ID) %>%
  dplyr::distinct()


human_common_sig_PCG <- human_sig %>%
  filter(gene_id %in% common_orthologs$Gene.stable.ID)
print(length(unique(human_common_sig_PCG$gene_id)))
human_common_sig_PCG <- human_common_sig_PCG %>%
  left_join(common_orthologs_filtered, by = c("gene_id" = "Gene.stable.ID"))
human_common_sig_PCG <- human_common_sig_PCG %>%
  dplyr::rename(ensembl_gene_id = Mouse.gene.stable.ID)

MS_common_sig_PCG <- MS_sig %>%
  filter(gene_id %in% common_orthologs$Mouse.gene.stable.ID)
print(length(unique(MS_common_sig_PCG$gene_id)))
MS_common_sig_PCG <- MS_common_sig_PCG %>%
  left_join(common_orthologs_filtered, by = c("gene_id" = "Mouse.gene.stable.ID"))
MS_common_sig_PCG <- MS_common_sig_PCG %>%
  dplyr::rename(ensembl_gene_id = Gene.stable.ID)

save(orthologs_filtered, human_common_sig_PCG, MS_common_sig_PCG, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/common_PCG.RData")
```

```{r combine Ulitsky, fish paper, and bioMart}
human_combined_sig <- bind_rows(human_common_sig_lncrna, human_common_sig, human_common_sig_PCG) %>%
  distinct()

MS_combined_sig <- bind_rows(MS_common_sig_lncrna, MS_common_sig, MS_common_sig_PCG) %>%
  distinct()

save(human_combined_sig, MS_combined_sig, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/combined_sig.RData")
```

```{r plotting tpm}
#for time model
# Load TPM data for TUG1_217
tpm_MS <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/MOUSE/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)

# Load TPM data for TUG1_217
tpm_human <- read.table("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/HUMAN/salmon.merged.gene_tpm.tsv", header=TRUE, row.names=1)
tpm_human <- tpm_human %>%
  dplyr::select(gene_name, gfp_0_1, gfp_0_2, gfp_0_3, gfp_12_1, gfp_12_2, gfp_12_3, gfp_24_1, gfp_24_2, gfp_24_3, gfp_48_1, gfp_48_2, gfp_48_3, gfp_96_1, gfp_96_2, gfp_96_3)

# Filter the TPM data for genes in sig_TUG1_217_LRT_shrnklfc
filtered_tpm_MS <- tpm_MS[tpm_MS$gene_name %in% MS_combined_sig$gene_name, ]
filtered_tpm_human <- tpm_human[tpm_human$gene_name %in% human_combined_sig$gene_name, ]

# Reshape mouse TPM data to long format
filtered_tpm_MS_long <- filtered_tpm_MS %>%
  pivot_longer(
    cols = starts_with("WT"),
    names_to = c("time_point", "replicate"),
    names_pattern = "^WT_(\\d+)_(\\d+)$",
    values_to = "tpm"
  ) %>%
  mutate(
    time_point = as.numeric(time_point),
    replicate = as.numeric(replicate),
    condition = "mouse"  # Add a column to label as TUG1_217
  )

# Reshape human TPM data to long format
filtered_tpm_human_long <- filtered_tpm_human %>%
  pivot_longer(
    cols = starts_with("gfp"),
    names_to = c("time_point", "replicate"),
    names_pattern = "^gfp_(\\d+)_(\\d+)$",
    values_to = "tpm"
  ) %>%
  mutate(
    time_point = as.numeric(time_point),
    replicate = as.numeric(replicate),
    condition = "human"  # Add a column to label as GFP
  )

# Combine human and mouse data into one data frame
tpm_combined_long <- bind_rows(filtered_tpm_MS_long, filtered_tpm_human_long)

# Calculate mean and standard error (SE) for TPM at each time point and condition
tpm_mean_combined <- tpm_combined_long %>%
  group_by(gene_name, time_point, condition) %>%
  summarise(
    mean_tpm = mean(tpm, na.rm = TRUE),
    se_tpm = sd(tpm, na.rm = TRUE) / sqrt(n()),  # Standard error
    .groups = 'drop'
  )

# Plot the combined TPM values for human and mouse with error bars
ggplot(tpm_mean_combined, aes(x = time_point, y = mean_tpm)) +
  geom_line(aes(color = condition, linetype = condition), alpha = 0.7) +  # Line colored by condition
  geom_point(aes(color = condition), alpha = 0.5) +  # Points for each mean TPM value
  geom_errorbar(
    aes(ymin = mean_tpm - se_tpm, ymax = mean_tpm + se_tpm, color = condition),
    width = 0.2, # Adjust the width of the error bars
    alpha = 0.7
  ) +
  facet_wrap(~ gene_name, scales = "free_y") +  # Separate plots for each gene
  labs(
    x = "Time Point (h)", 
    y = "Mean TPM", 
    color = "Condition"
  ) +
  theme_minimal() +  # Minimal theme
  scale_color_manual(values = c("human" = "darkred", "mouse" = "blue")) +  # Manual color scale
  scale_linetype_manual(values = c("human" = "solid", "mouse" = "dashed")) + 
  scale_x_continuous(breaks = unique(tpm_mean_combined$time_point)) +  # Set x-axis breaks
  theme(
    aspect.ratio = 0.6,  # Aspect ratio of 3:2 
    strip.text = element_text(size = 12),  # Adjust facet labels font size if necessary
  )

# Save the plot
ggsave(
  path = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MOUSEVSHUMAN/ML/figures/",
  filename = "combined_sig_TPM_individule_with_errorbars.pdf",
  height = 20, width = 30, limitsize = FALSE
)
```

```{r plotting log2FC}
# Add a 0 time_point for mouse data
MS_combined_sig_zero <- MS_combined_sig %>%
  dplyr::select(gene_id, gene_name, baseMean) %>%
  distinct() %>%
  mutate(log2FoldChange = 0,
         timepoint = 0,
         source = "mouse")

MS_combined_sig <- MS_combined_sig %>%
  bind_rows(MS_combined_sig_zero)

# Prepare TUG1_217 data with source column
MS_combined_sig <- MS_combined_sig %>%
  mutate(source = "TUG1_217") %>%
  

# Add a 0 time_point for TUG1_217
TUG1_217_trend_sig_filtered_zero <- TUG1_217_trend_sig_filtered %>%
  dplyr::select(gene_id, gene_name, baseMean) %>%
  distinct() %>%
  mutate(log2FoldChange = 0,
         timepoint = 0,
         source = "TUG1_217")

TUG1_217_trend_sig_filtered <- TUG1_217_trend_sig_filtered %>%
  bind_rows(TUG1_217_trend_sig_filtered_zero)

# Combine TUG1_217 and filtered GFP data
combined_data <- bind_rows(TUG1_217_trend_sig_filtered, filtered_GFP_res)

# Generate the plot
ggplot(combined_data, aes(x = timepoint, y = log2FoldChange, group = interaction(gene_id, source), color = source)) +
  geom_hline(yintercept = 0, linetype = "dashed") +
  geom_line(alpha = 0.7, aes(linetype = source)) +
  geom_point(alpha = 0.8) +
  facet_wrap(~gene_name, scales = "free_y") +
  scale_color_manual(values = c("TUG1_217" = "darkred", "GFP" = "blue")) +
  scale_linetype_manual(values = c("TUG1_217" = "solid", "GFP" = "dashed")) + 
  scale_x_continuous(breaks = c(0, 2, 4, 8, 16, 24, 48, 96), labels = c("0", "2", "4", "8", "16", "24", "48", "96")) +
  theme_minimal() +
  labs(title = "Gene LFC Trends for TUG1_217 and GFP",
       x = "Timepoint (hours)",
       y = "Log2 Fold Change",
       color = "Source",
       linetype = "Source") +
   theme(
    aspect.ratio = 0.6,  # Aspect ratio of 3:2 
    strip.text = element_text(size = 12),  # Adjust facet labels font size if necessary
  )

# Save the plot
ggsave(path = "/scratch/Shares/rinn/ML/RNAseq_FULL_MODEL/TUG1_217/figures/", 
       filename = "TUG1_217_vs_GFP_trend_sig_lfc_individule.pdf", 
       height = 60, width = 80, limitsize = FALSE)


```


```{r GO}
library(topGO)
library(org.Hs.eg.db)

# Create GO data object
GOdata <- new("topGOdata", 
              ontology = "BP", 
              allGenes = gene_list, 
              geneSel = gene_selection, 
              annot = annFUN.org, 
              mapping = "org.Hs.eg.db", 
              ID = "SYMBOL")

# Run Fisherâ€™s exact test
result <- runTest(GOdata, algorithm = "weight01", statistic = "fisher")

# Get significant terms
allRes <- GenTable(GOdata, raw.p.value = result, topNodes = 10)
print(allRes)


# Combine results for plotting
go_up@result$regulation <- "Upregulated"
go_down@result$regulation <- "Downregulated"
combined_go <- rbind(go_up@result, go_down@result)

library(ggplot2)
ggplot(combined_go, aes(x = -log10(p.adjust), y = Description, color = regulation)) +
    geom_point(aes(size = Count)) +
    labs(title = "GO Enrichment Analysis (Up vs. Downregulated Genes)",
         x = "-log10 Adjusted P-value",
         y = "GO Term") +
    scale_color_manual(values = c("Upregulated" = "red", "Downregulated" = "blue")) +
    theme_minimal()

```

