---
output: html_document
editor_options: 
  chunk_output_type: console
---
```{r}
#BiocManager::install("biomaRt")
library(biomaRt)
```

```{r}
# Use the Ensembl mirror
#human_mart <- useEnsembl(biomart = "ensembl", dataset = "hsapiens_gene_ensembl", mirror = "www")
#mouse_mart <- useEnsembl(biomart = "ensembl", dataset = "mmusculus_gene_ensembl", mirror = "www")

# Retrieve orthologs
#orthologs <- getLDS(attributes = c("ensembl_gene_id", "external_gene_name"),
                    #filters = "ensembl_gene_id",
                    #values = NULL, # Specify values if needed
                    #mart = human_mart,
                    #attributesL = c("ensembl_gene_id", "external_gene_name"),
                    #martL = mouse_mart)
###!!Error: biomaRt has encountered an unexpected server error.!!###
```

```{r}
# download orthologs info from https://www.ensembl.org/biomart/martview
# about how to download, refer https://www.ensembl.org/info/data/biomart/how_to_use_biomart.html'

# load to R
orthologs <- read.csv("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/mart_export.txt", header = TRUE, stringsAsFactors = FALSE, sep = ",")
synteny <- read.delim("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/human_mouse_syntenic_lncRNA.txt",
                      header = TRUE, stringsAsFactors = FALSE, sep = "\t")
gencode_synteny <- read.delim("/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/DATA/GENCODEvGENCODE_orthologs_v3.txt",
                      header = TRUE, stringsAsFactors = FALSE, sep = "\t")

# View the first few rows
head(orthologs)
colnames(orthologs)
head(synteny)

orthologs_filtered <- orthologs[orthologs$Mouse.homology.type == "ortholog_one2one", ]

save(orthologs, orthologs_filtered, file = "/scratch/Shares/rinnclass/MASTER_CLASS/STUDENTS/genehomies/RESULTS/MS_vs_HS/orthologs.RData")

#now, we are ready to find genes are changed in both human and mouse stem cells
```

